### DISCUSSION TITLE
NoSQL (Document) Database Alternative?

### DISCUSSION THREAD

**Discussion by reinhart1010** [supabase/supabase#769] [Feature Requests] [answered]
Since Firebase itself offers certain types of NoSQL databases (Realtime Database, Cloud Firestore) it would be great for Supabase to include a NoSQL alternative, especially document database systems. I have used Firebase Realtime Database, Apache CouchDB and mongoDB at some of my past projects.

Firebase Realtime Database has a unique data structure (e.g. no regular arrays, each database is a large Document) which makes it unique than regular document-based databases such as mongoDB, where each mongoDB "database" (term borrowed from SQL and relational databases) is considered a Collection of Document. Making an open source, drop-in replacement for it would take time and some security considerations.

However it would be possible to include a regular document-based NoSQL database such as [Apache CouchDB](https://couchdb.apache.org) into Supabase, so any Supabase projects which explicitly needs NoSQL can happily use the service. Several CouchDB features such as [mango](https://github.com/cloudant/mango) allows mongoDB queries to run on CouchDB databases.

**Comment by pmauchle** [comment, marked as answer]
Postgresql has a jsonb datatype where you can store documents like in NoSQL database and query them. https://supabase.io/docs/client/select#quering-json-data

**Reply by reinhart1010** [reply]
Oh, never thought of PostgreSQL's JSON functionality before. So in order to fully replicate a new "Collection" of "Document" (terms borrowed from mongoDB)  I'll only need to create a single-column table on it?

**Reply by steve-chavez** [reply]
@reinhart1010 Yes, you can do that. However, you might want to add a primary key in addition to the JSON column so you can JOIN different collections(actually relations in this case) through [supabase-js](https://supabase.io/docs/client/select#query-foreign-tables).

I recommend you check this tutorial to get acquainted with JSON in PostgreSQL: https://www.postgresqltutorial.com/postgresql-json/
